### DISCUSSION TITLE
how to improve data insert performance on milvus 2.0

### DISCUSSION THREAD

**Discussion by shawn-gogh** [milvus-io/milvus#17835] [Q&A and General discussion]
Hi!

The [benchmark page](https://milvus.io/docs/v2.0.x/benchmark.md) list the data inserting performance on  a single datanode.but i'm wondering, can we  improve the inserting performance by:

1. increase the replicas number of datanode?
2. increase the resource of  memory and cpu?

with this questions  i tried but failed. I found that the inserting performance still the same dispite the increasing of datanode number and resource.

why and how?

Milvus version: 2.0.2

**Comment by yhmo** [comment]
When you create a collection, there is a parameter named "shard_num":
```
class Collection:
    """ This is a class corresponding to collection in milvus. """
    def __init__(self, name, schema=None, using="default", shards_num=2, **kwargs):
```
The default value is 2. This parameter controls how many "channels" consume the data.
Assume you create a collection with shard_num=2, milvus will create two data channels for this collection. Then you insert data, let's say you insert 10000 entities. Milvus will split then into two parts, each part has about 5000 entities, then send the two parts to two data nodes to process parallelly. If the shard_num=4, then there will be four parts processed parallelly. 

So, increase data node number cannot improve insert performance. You can try shard_num=4, the performance would be better.

**Reply by shawn-gogh** [reply]
@yhmo Thank you for your reply. I'm really appreciate it!

But i'm wondering... if the larger the shard_num is the faster the inserting action is, and  if it is the truth, we can just set the number very large to make the inserting faster!
And i also want to know:

1.  the relationship between the parameter shard_num and the inserting performance, is it linear?
2.  if there exist other method to improve the inserting performation except the shard_num parameter?

Looking forward to your reply!

**Reply by shawn-gogh** [reply]
I tried to set the shard_num to 4 and 8 ,but the 100k inserting time is still the same 😞

**Reply by yhmo** [reply]
I just test shard_num=1,2,4,8 on one machine with a single data node, the insert performance is best when shard_num=1, worst when shard_num=8.

Ok, I am wrong on this "increase data node number cannot improve insert performance", because if there is only one data node to process multiple parts, multi-threads may contest on resources including cpu/memory/storage.

Ideally, if the shard num is 4, it is better to prepare 4 data nodes on different hosts, each node processing one part.

**Reply by yhmo** [reply]
I guess "the insert performance is best when shard_num is equal to data node count", but I don't get time to practice. You can have a try.

**Reply by shawn-gogh** [reply]
Got it!  thank you once again,you are very helpful, i'll try it later.

**Reply by hjlee9182** [reply]
@shawn-gogh 
Hello, Can you share your experiment result?
